common/src/node.rs

```diff
@@ -93,6 +93,52 @@ impl VoxelData {
     }
 }
 
+/// Contains the context needed to know the locations of individual cubes within a chunk in the chunk's coordinate
+/// system. A given `ChunkLayout` is uniquely determined by its dimension.
+pub struct ChunkLayout {
+    dimension: usize,
+    dual_to_grid_factor: f32,
+}
+
+impl ChunkLayout {
+    pub fn new(dimension: usize) -> Self {
+        ChunkLayout {
+            dimension,
+            dual_to_grid_factor: Vertex::dual_to_chunk_factor() as f32 * dimension as f32,
+        }
+    }
+
+    /// Number of cubes on one axis of the chunk. Margins are not included.
+    pub fn dimension(&self) -> usize {
+        self.dimension
+    }
+
+    /// Scale by this to convert dual coordinates to homogeneous grid coordinates.
+    pub fn dual_to_grid_factor(&self) -> f32 {
+        self.dual_to_grid_factor
+    }
+
+    /// Converts a single coordinate from dual coordinates in the Klein-Beltrami model to an integer coordinate
+    /// suitable for voxel lookup. Margins are included. Returns `None` if the coordinate is outside the chunk.
+    #[inline]
```

Comment:
If we're gunna bother to mark things inline, definitely do so for the pure getters above.